### DISCUSSION TITLE
How to do complex aggregations

### DISCUSSION THREAD

**Discussion by qqlearn123** [duckdb/duckdb#8241] [Q&A]
I have a dataframe in Python and I want to do a groupby aggregation on it. I have attempted the following:
```
import duckdb
import pandas as pd

df = pd.DataFrame({
    "id": [1, 1, 1, 2, 2, 2, 3, 3, 3],
    "id2": [1, 1, 1, 1, 1, 1, 2, 2, 2],
    "v": [1, 2, 3, 4, 5, 6, 7, 8, 9],
    "v2": [9, 8, 7, 6, 5, 4, 3, 2, 1]
})

res = duckdb.sql(
    """
    SELECT id, id2, sum((v - mean(v)) / sum(abs(v - mean(v))) * v2)
    FROM df
    GROUP BY id, id2
    """
).pd()
```

But I got a binding error: 'aggregate function calls cannot be nested'.

What can I do to make such an aggregation works? Either through SQL with a different syntax or is there some other more pandas way to do such a thing using duckdb?

**Comment by orzom411** [comment]
```
import duckdb
import pandas as pd

df = pd.DataFrame({
    "id": [1, 1, 1, 2, 2, 2, 3, 3, 3],
    "id2": [1, 1, 1, 1, 1, 1, 2, 2, 2],
    "v": [1, 2, 3, 4, 5, 6, 7, 8, 9]
})


res = duckdb.sql(
    """
    with x as (
    SELECT d.id, d.id2, d.v, d.v - mean(d.v) over( partition by null ) as m
    FROM df as d )
    select x.id, x.id2, mean(x.v) / sum(abs(x.m)) as s
    from x as x 
    group by x.id, x.id2;    
    """ );

res;
```

Not sure if this is the result you expected:
```
┌───────┬───────┬────────────────────┐
│  id   │  id2  │         s          │
│ int64 │ int64 │       double       │
├───────┼───────┼────────────────────┤
│     1 │     1 │ 0.2222222222222222 │
│     2 │     1 │                2.5 │
│     3 │     2 │ 0.8888888888888888 │
└───────┴───────┴────────────────────┘
```

Few adjustments, 
1) wasn't sure about your pandas import, so I did it as pd.
2) Notice employing a Common Table Expression as the error noted can't perform a nest mean within a sum function.  Also "v" wasn't included in in the group by or aggregate.

Hope this helps.

edit: missed a few aliases (_try to always alias in SQL, it mutes much confusion down the road_).

**Reply by qqlearn123** [reply]
Thanks for your response. I was missing another variable and calculation step in my original post. I just added it.

I think in the CTE I will need to do  'partition by id, id2' as well.

Does this kind of aggregation have to be done via CTE? Since, with CTE, it will involve both a grouping operation and a window operation. If everything can be done in group by, the window operation could be saved, so that no need to do splitting twice.

**Comment by orzom411** [comment]
Generally it helps if if the desired result is provided.  Maybe able to avoid the window entirely, that is if the result correct?

**Reply by qqlearn123** [reply]
The desired result will be
```
id id2 res
1   1   -1
2   1   -1
3   2   -1
```

**Reply by orzom411** [reply]
Good Morning,

Looking over your results I'm not sure how your are arriving at -1 result.  Specifically ABS function function returns a positive value (https://duckdb.org/docs/sql/functions/numeric.html).

The short of the long, I try to start with a simple result set and proof of possible.  If I don't understand the results I go back to the business unit and ask them to detail out how they arrived at the given result.

I wish you a truly wonderful week.

**Reply by qqlearn123** [reply]
Hello, the result contains a division where the numerator is v - mean(v), which can be negative. I think what you mean positive is the denominator part.

**Comment by orzom411** [comment]
Sure, but the example dataset you've provided had positive values, 1 through 9.  Mean(v), if V is always positive as is your example dataset, I don't think can be negative.

Sorry I wasn't able to help any further.

**Reply by qqlearn123** [reply]
It is actually `v - mean(v)`, for example, if `v` is 1, 2, 3 in a group, `v - mean(v)` will be -1, 0, 1. So, it can be negative.

**Comment by orzom411** [comment]
Okay, I'm still confused about your data as you have v between 1 and 9 and the addition of reverse scale.  

Maybe try the sql sign function, it gives you -1, 0 or 1, maybe it simplifies the math here?

Nesting aggregate isn't mathematically sound, e.g. what are you trying to sum a sum, but what is that sum supposed to be?  If you can break it down into a full truth table, all four values, and what the result is expected to be you might be able reduce the math to just a simple valid formulation.

Wishing you luck and if you do solve it you might want to post to this thread your solution so that someone landing on nested summation can see your progress.

**Reply by qqlearn123** [reply]
Let me try to do a walkthrough calculation for the first group (id=1 and id2=1 group) just to clarify:
- `v - mean(v)` as `p1`: -1, 0, 1
- `sum(abs(v - mean(v)))` as `p2`: 1 + 0 + 1 = 2
- `v2`: 9, 8, 7
- `result`: `sum(p1 / p2 * v2)` = -1 * 9 + 0 * 8 + 1 * 7 / 2 = -1

The nested aggregation happens in the `p2` step - sum of absolute demean values. And, I think it is a reasonable need.

**Comment by orzom411** [comment]
Maybe you could say it a little differently, like, and correct me if I am incorrect:

- v - mean(v) as p1: -1, 0, 1

- sum(abs(p1)) as p2: 1 + 0 + 1 = 2

- v2:` 9, 8, 7

- result: sum(p1 / p2 * v2) = -1 * 9 + 0 * 8 + 1 * 7 / 2 = -1

And that kind of tells you your cte steps.
